src/librepgp/repgp.c

```diff
@@ -0,0 +1,362 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <errno.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#include <rnp/rnp_def.h>
+#include <rnp/rnp.h>
+#include <rnp/rnpcfg.h>
+#include <repgp/repgp.h>
+#include <rekey/rnp_key_store.h>
+
+#include "internal_types.h"
+#include "packet-print.h"
+#include "memory.h"
+#include "utils.h"
+#include "crypto.h"
+#include "reader.h"
+#include "validate.h"
+
+repgp_handle_t *
+create_filepath_handle(const char *filename)
+{
+    if (!filename) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    repgp_handle_t *s = calloc(sizeof(*s), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->filepath = strndup(filename, strlen(filename));
+    s->type = REPGP_HANDLE_FILE;
+    return s;
+}
+
+repgp_handle_t *
+create_buffer_handle(const size_t buffer_size)
+{
+    repgp_handle_t *s = calloc(sizeof(*s), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->buffer.data = (unsigned char *) malloc(buffer_size);
+    if (!s->buffer.data) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->buffer.size = buffer_size;
+    s->type = REPGP_HANDLE_BUFFER;
+    return s;
+}
+
+repgp_handle_t *
+create_data_handle(const uint8_t *data, size_t data_len)
+{
+    repgp_handle_t *s = calloc(sizeof(*s), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    s->buffer.data = (unsigned char *) malloc(data_len);
+    if (!s->buffer.data) {
+        return REPGP_HANDLE_NULL;
+    }
+    memcpy(s->buffer.data, data, data_len);
+
+    s->buffer.size = data_len;
+    s->type = REPGP_HANDLE_BUFFER;
+    return s;
+}
+
+/* Reads into memory everything from stdin */
+repgp_handle_t *
+create_stdin_handle(void)
+{
+    char     buf[BUFSIZ * 8];
+    uint8_t *data = NULL;
+    size_t   size = 0;
+    ssize_t  n;
+
+    repgp_handle_t *s = calloc(sizeof(*s), 1);
+    if (!s) {
+        return REPGP_HANDLE_NULL;
+    }
+
+    /* Read in everything and keeps it in memory.
+     * For stdin it kind of makes sense as no one
+     * should provide a lot of data on stdin.
+     */
```

Comment:
This doesn't really have to be fixed in this PR but IMO this wouldn't be acceptable.
Many users will not think twice about doing something like `rnp --decrypt < enormousfile > decrypted` because typically *nix utils are really good at this sort of thing.

Reply:
Agree with @dewyatt. Maybe @ni4's work in #238 could help with this?

Reply:
I agree here. But this problem already exist in ``rnp``, it's neither getting introduced nor fixed by this PR.

Reply:
@flowher could we just leave a TODO comment in this code that we need to fix this for handling large files? I'm not sure if we already have an issue for this, but if not we should probably make one.

Reply:
TODO added.

Reply:
Thanks @flowher !

Reply:
Since @ni4 will handle this, let's move on. 👏